#### ONE-HOT-ENCODING

OHE(One-Hot-Encoding),Consist of replacing categorical variable with boolean variable(0,1)

Each one of the boolean varaible is also known as DUMMY VARIABLE/BINARY VARIABLE

For example, from the categorical variable "Gender", with labels 'female' and 'male', we can generate the boolean variable "female", which takes 1 if the person is female or 0 otherwise. We can also generate the variable male, which takes 1 if the person is "male" and 0 otherwise.

In [2]:
import pandas as pd

data = pd.read_csv('/Users/nandrajog/Downloads/titanic.csv', usecols=['Sex'])
data.head()

Sex
0    male
1  female
2  female
3  female
4    male

In [3]:
pd.get_dummies(data)

Sex_female  Sex_male
0             0         1
1             1         0
2             1         0
3             1         0
4             0         1
..          ...       ...
886           0         1
887           1         0
888           1         0
889           0         1
890           0         1

[891 rows x 2 columns]

In [4]:
pd.concat([data,pd.get_dummies(data)],axis=1)

Sex  Sex_female  Sex_male
0      male           0         1
1    female           1         0
2    female           1         0
3    female           1         0
4      male           0         1
..      ...         ...       ...
886    male           0         1
887  female           1         0
888  female           1         0
889    male           0         1
890    male           0         1

[891 rows x 3 columns]

As you may have noticed, we only need 1 of the 2 dummy variables to represent the original categorical variable Sex. Any of the 2 will suffice, and it doesn't matter which one we select, since they are equivalent.

Therefore, to encode a categorical variable with 2 labels, we need 1 dummy variable.

To extend this concept, to encode categorical variable with k labels, we need k-1 dummy variables.

How can we get this using pandas?

In [5]:
# obtaining k-1 labels
pd.get_dummies(data, drop_first=True).head()

Sex_male
0         1
1         0
2         0
3         0
4         1

In [6]:
# Let's now look at an example with more than 2 labels

data=pd.read_csv('/Users/nandrajog/Downloads/titanic.csv', usecols=['Embarked'])
data.head()

Embarked
0        S
1        C
2        S
3        S
4        S

In [7]:
data.Embarked.unique()

array(['S', 'C', 'Q', nan], dtype=object)

In [8]:
# get whole set of dummy variables

pd.get_dummies(data).head()

Embarked_C  Embarked_Q  Embarked_S
0           0           0           1
1           1           0           0
2           0           0           1
3           0           0           1
4           0           0           1

In [10]:
pd.get_dummies(data,drop_first=True)

Embarked_Q  Embarked_S
0             0           1
1             0           0
2             0           1
3             0           1
4             0           1
..          ...         ...
886           0           1
887           0           1
888           0           1
889           0           0
890           1           0

[891 rows x 2 columns]

In [12]:
# This below code will create extra col for NAN

# 1) Drop First

# 2) Add NAN

pd.get_dummies(data, drop_first=True, dummy_na=True).head()

Embarked_Q  Embarked_S  Embarked_nan
0           0           1             0
1           0           0             0
2           0           1             0
3           0           1             0
4           0           1             0

#### Notes
Both pandas and sklearn will provide a whole set of dummy variables from a categorical variable. This is, instead of returning k-1 binary variables, they will return k, with the option in pandas of dropping the first binary variable and obtain k-1.

### When should you use k and when k-1?

#### One hot encoding into k-1

One hot encoding into k-1 binary variables takes into account that we can use 1 less dimension and still represent the whole information: if the observation is 0 in all the binary variables, then it must be 1 in the final (removed) binary variable. As an example, for the variable gender encoded into male, if the observation is 0, then it has to be female. We do not need the additional female variable to explain that.

And the same is true for all machine learning algorithms that look at ALL the features at the same time during training. For example, support vector machines and neural networks as well. And clustering algorithms.

#### One hot encoding into k dummy variables

However, tree based models select at each iteration only a group of features to make a decision. This is to separate the data at each node.
Therefore K binary variable should be used

#### Finally, if you are planning to do feature selection, you will also need the entire set of binary variables (k) to let the machine learning model select which ones have the most predictive power.

In [13]:
import numpy as np

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import roc_auc_score
from sklearn.model_selection import train_test_split

In [14]:
data = pd.read_csv('/Users/nandrajog/Downloads/titanic.csv',low_memory=False)
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [15]:
data_OHE = pd.concat([data[['Pclass', 'Age', 'SibSp','Parch', 'Survived']], # numerical variables 
                      pd.get_dummies(data.Sex, drop_first=True),   # binary categorical variable
                      pd.get_dummies(data.Embarked, drop_first=True)],  # k categories in categorical
                    axis=1)

data_OHE.head()

Pclass   Age  SibSp  Parch  Survived  male  Q  S
0       3  22.0      1      0         0     1  0  1
1       1  38.0      1      0         1     0  0  0
2       3  26.0      0      0         1     0  0  1
3       1  35.0      1      0         1     0  0  1
4       3  35.0      0      0         0     1  0  1

In [16]:
# and now let's separate into train and test set

X_train, X_test, y_train, y_test = train_test_split(data_OHE[['Pclass', 'Age', 'SibSp',
                                                              'Parch', 'male', 'Q', 'S']].fillna(0),
                                                    data_OHE.Survived,
                                                    test_size=0.3,
                                                    random_state=0)
X_train.shape, X_test.shape

((623, 7), (268, 7))

In [18]:
# let's build a random forest

rf = RandomForestClassifier(n_estimators=200, random_state=39, max_depth=3)
rf.fit(X_train, y_train)
print('Train set')
pred = rf.predict_proba(X_train)
print('Random Forests roc-auc: {}'.format(roc_auc_score(y_train, pred[:,1])))
print('Test set')
pred = rf.predict_proba(X_test)
print('Random Forests roc-auc: {}'.format(roc_auc_score(y_test, pred[:,1])))


Train set
Random Forests roc-auc: 0.8696177957094207
Test set
Random Forests roc-auc: 0.8606250000000001


In [19]:
# and now an AdaBoost classifier

ada = AdaBoostClassifier(n_estimators=200, random_state=44)
ada.fit(X_train, y_train)
print('Train set')
pred = ada.predict_proba(X_train)
print('AdaBoost roc-auc: {}'.format(roc_auc_score(y_train, pred[:,1])))
print('Test set')
pred = ada.predict_proba(X_test)
print('AdaBoost roc-auc: {}'.format(roc_auc_score(y_test, pred[:,1])))

Train set
AdaBoost roc-auc: 0.9010433613153728
Test set
AdaBoost roc-auc: 0.8474404761904762


In [21]:
# and finally a logistic regression

logit = LogisticRegression(random_state=44)
logit.fit(X_train, y_train)
print('Train set')
pred = logit.predict_proba(X_train)
print('Logistic Regression roc-auc: {}'.format(roc_auc_score(y_train, pred[:,1])))
print('Test set')
pred = logit.predict_proba(X_test)
print('Logistic Regression roc-auc: {}'.format(roc_auc_score(y_test, pred[:,1])))

Train set
Logistic Regression roc-auc: 0.8452745059760092
Test set
Logistic Regression roc-auc: 0.8424404761904761


/anaconda3/envs/C-vision/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Logistic regression and Adaboost show the highest performance using the current datasets.

### One Hot encoding: final notes¶

If the variable has loads of categories, then OHE increases the feature space dramatically

In [30]:
#let's load the variable Cabin of the titanic dataset

data = pd.read_csv('/Users/nandrajog/Downloads/titanic.csv', usecols = ['Cabin'])
data.head()

Cabin
0   NaN
1   C85
2   NaN
3  C123
4   NaN

In [31]:
data.Cabin

0       NaN
1       C85
2       NaN
3      C123
4       NaN
       ... 
886     NaN
887     B42
888     NaN
889    C148
890     NaN
Name: Cabin, Length: 891, dtype: object

In [28]:
data.Cabin.unique()

array([nan, 'C85', 'C123', 'E46', 'G6', 'C103', 'D56', 'A6',
       'C23 C25 C27', 'B78', 'D33', 'B30', 'C52', 'B28', 'C83', 'F33',
       'F G73', 'E31', 'A5', 'D10 D12', 'D26', 'C110', 'B58 B60', 'E101',
       'F E69', 'D47', 'B86', 'F2', 'C2', 'E33', 'B19', 'A7', 'C49', 'F4',
       'A32', 'B4', 'B80', 'A31', 'D36', 'D15', 'C93', 'C78', 'D35',
       'C87', 'B77', 'E67', 'B94', 'C125', 'C99', 'C118', 'D7', 'A19',
       'B49', 'D', 'C22 C26', 'C106', 'C65', 'E36', 'C54',
       'B57 B59 B63 B66', 'C7', 'E34', 'C32', 'B18', 'C124', 'C91', 'E40',
       'T', 'C128', 'D37', 'B35', 'E50', 'C82', 'B96 B98', 'E10', 'E44',
       'A34', 'C104', 'C111', 'C92', 'E38', 'D21', 'E12', 'E63', 'A14',
       'B37', 'C30', 'D20', 'B79', 'E25', 'D46', 'B73', 'C95', 'B38',
       'B39', 'B22', 'C86', 'C70', 'A16', 'C101', 'C68', 'A10', 'E68',
       'B41', 'A20', 'D19', 'D50', 'D9', 'A23', 'B50', 'A26', 'D48',
       'E58', 'C126', 'B71', 'B51 B53 B55', 'D49', 'B5', 'B20', 'F G63',
       'C62 C64',

In [29]:
Cabin_OHE = pd.get_dummies(data.Cabin)
Cabin_OHE.shape

(891, 147)

In [25]:
Cabin_OHE

A10  A14  A16  A19  A20  A23  A24  A26  A31  A32  ...  E8  F E69  F G63  \
0      0    0    0    0    0    0    0    0    0    0  ...   0      0      0   
1      0    0    0    0    0    0    0    0    0    0  ...   0      0      0   
2      0    0    0    0    0    0    0    0    0    0  ...   0      0      0   
3      0    0    0    0    0    0    0    0    0    0  ...   0      0      0   
4      0    0    0    0    0    0    0    0    0    0  ...   0      0      0   
..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ..    ...    ...   
886    0    0    0    0    0    0    0    0    0    0  ...   0      0      0   
887    0    0    0    0    0    0    0    0    0    0  ...   0      0      0   
888    0    0    0    0    0    0    0    0    0    0  ...   0      0      0   
889    0    0    0    0    0    0    0    0    0    0  ...   0      0      0   
890    0    0    0    0    0    0    0    0    0    0  ...   0      0      0   

     F G73  F2  F33  F38  F4  G6  T  
0        0   0    0    0   0   0  0  
1        0   0    0    0   0   0  0  
2        0   0    0    0   0   0  0  
3        0   0    0    0   0   0  0  
4        0   0    0    0   0   0  0  
..     ...  ..  ...  ...  ..  .. ..  
886      0   0    0    0   0   0  0  
887      0   0    0    0   0   0  0  
888      0   0    0    0   0   0  0  
889      0   0    0    0   0   0  0  
890      0   0    0    0   0   0  0  

[891 rows x 147 columns]

If we performed OHE in the variable Cabin that contains 148 different labels, we would end up with 147 variables, where originally there was one. If we have a few categorical variables like this, we would end up with huge datasets. Therefore, OHE is not always the best option to encode categorical variables.